runtime/src/main/jni/Runtime.cpp

```diff
@@ -301,9 +309,39 @@ jint Runtime::GenerateNewObjectId(JNIEnv *env, jobject obj)
 	return objectId;
 }
 
-void Runtime::AdjustAmountOfExternalAllocatedMemoryNative(JNIEnv *env, jobject obj, jlong usedMemory)
+void Runtime::AdjustAmountOfExternalAllocatedMemory() {
+	int64_t usedMemory = m_env.CallLongMethod(m_runtime, GET_USED_MEMORY_METHOD_ID);
+	int64_t changeInBytes = usedMemory - m_lastUsedMemory;
+	int64_t externalMemory = 0;
+
+	if (changeInBytes != 0) {
+		externalMemory = m_isolate->AdjustAmountOfExternalAllocatedMemory(changeInBytes);
+	}
+
+	DEBUG_WRITE("usedMemory=%" PRId64 " changeInBytes=%" PRId64 " externalMemory=%" PRId64, usedMemory, changeInBytes, externalMemory);
+
+	m_lastUsedMemory = usedMemory;
+}
+
+bool Runtime::NotifyGC(JNIEnv *env, jobject obj)
 {
-	m_isolate->AdjustAmountOfExternalAllocatedMemory(usedMemory);
+	bool success = __sync_bool_compare_and_swap(&m_runGC, false, true);
+	return success;
+}
+
+bool Runtime::TryCallGC() {
```

Comment:
After internal discussion it seems we are left with one more variant to test before resorting to full gc the current PR way. 

As for the current PR approach maybe we should remove the JS transition call and call the Isolate::RequestGarbageCollectionForTesting directly